### PR TITLE
Voronin

### PR DESCRIPTION
**Pull request by AARaven**
Hello, could you please add my changes into your main branch

### CODE REVIEW

File: src/main/java/provider/WebDriverProvider.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package provider;
```
Comment (a-klimov) [inline src/main/java/provider/WebDriverProvider.java]: Please use group id of the project for packages names: com.sperasoft.edu.

File: src/main/java/utils/FrameworkUtils.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package utils;
```
Comment (a-klimov) [inline src/main/java/utils/FrameworkUtils.java]: Please use group id of the project for packages names: com.sperasoft.edu.

File: src/main/java/web/page/AbstractWebPage.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package web.page;
```
Comment (a-klimov) [inline src/main/java/web/page/AbstractWebPage.java]: Please use group id of the project for packages names: com.sperasoft.edu.

File: src/main/java/web/page/AbstractWebPage.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Getter @Setter
+    private URL url;
+
+    @Getter ( AccessLevel.PROTECTED ) @Setter ( AccessLevel.PRIVATE )
+    private WebDriver driver;
+
+    @SneakyThrows
+    public AbstractWebPage( WebDriver driver ) {
+        setUrl( new URL( getProperties( getClass().getSimpleName() ,
```
Comment (a-klimov) [inline src/main/java/web/page/AbstractWebPage.java]: Please use **this** keyword

File: src/test/java/com/sperasoft/edu/bdd/pages/BasePage.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+                 @FindBy ( css = "div.name" ) } )
+    private WebElement textName;
+
+    public void verifyUserName( String name ) {
+        Assert.assertTrue( getTextName().isDisplayed() );
+        Assert.assertEquals( name ,
+                             getTextName().getText() );
+    }
+
+    public BasePage( WebDriver driver ) {
```
Comment (a-klimov) [inline src/test/java/com/sperasoft/edu/bdd/pages/BasePage.java]: This is redundant

File: src/test/resources/features/UserLogin.feature [hunk 6]
```diff
@@ -0,0 +1,30 @@
+########################################################################################################################
+
+Feature: Login functionality
+    
+    @positive
+    Scenario: Verifying that user can login to lostfilm.tv with own username and password
+        
+        Given I open the login page
+        When  I type email and password to input areas
```
Comment (a-klimov) [inline src/test/resources/features/UserLogin.feature]: 'input areas' don't sound friendly for everyone

File: src/test/resources/features/UserLogin.feature [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            | password | Mzaq1xsw2                  |
+        And   I click on submit button
+        Then  I successfully login
+        And   My name is correctly displayed
+            | name | Testing |
+    
+    @negative
+    Scenario: Verifying that user can't login to lostfilm.tv with wrong username and password
+        
+        Given I open the login page
```
Comment (a-klimov) [inline src/test/resources/features/UserLogin.feature]: talking about the future. If we are going to open other pages, we will need to create new step with a difference in one word, annd add an additional implementation

File: src/test/resources/features/UserLogin.feature [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        When  I type email and password to input areas
+            | email    | example@gmail.com |
+            | password | 123345            |
+        And   I click on submit button
```
Comment (a-klimov) [inline src/test/resources/features/UserLogin.feature]: talking about the future. If we are going to click other buttons, we will need to create new step with a difference in one word, and add an additional implementation

File: src/test/resources/features/UserLogin.feature (hunk 6 above)
Comment (a-klimov) [inline src/test/resources/features/UserLogin.feature]: Word 'and' in scenarios means that a step is complicated. Please use "I enter accounts credentials" or split into 2 steps

File: src/test/resources/features/UserLogin.feature [hunk 9]
```diff
@@ -0,0 +1,30 @@
+########################################################################################################################
+
+Feature: Login functionality
+    
+    @positive
+    Scenario: Verifying that user can login to lostfilm.tv with own username and password
```
Comment (a-klimov) [inline src/test/resources/features/UserLogin.feature]: Scenario name should be a statement "User can ...."

File: src/main/java/utils/FrameworkUtils.java (hunk 2 above)
Comment (AARaven) [reply src/main/java/utils/FrameworkUtils.java]:
```suggestion
change directory -> src/main/java/com/sperasoft/edu/bdd/provider/WebDriverProvider.java
```

File: src/main/java/web/page/AbstractWebPage.java (hunk 3 above)
Comment (AARaven) [reply src/main/java/web/page/AbstractWebPage.java]:
```suggestion
package com.sperasoft.edu.bdd.web.page
```

File: src/main/java/provider/WebDriverProvider.java (hunk 1 above)
Comment (AARaven) [reply src/main/java/provider/WebDriverProvider.java]:
```suggestion
package com.sperasoft.edu.bdd.provider;
```

File: src/main/java/web/page/AbstractWebPage.java (hunk 3 above)
Comment (AARaven) [reply src/main/java/web/page/AbstractWebPage.java]:
```suggestion
package com.sperasoft.edu.bdd.web.page;
```

File: src/main/java/web/page/AbstractWebPage.java (hunk 4 above)
Comment (AARaven) [reply src/main/java/web/page/AbstractWebPage.java]:
```suggestion
        @SneakyThrows
    public AbstractWebPage( WebDriver driver ) {
       this.url = new URL( getProperties( getClass().getSimpleName() ,
```

File: src/test/java/com/sperasoft/edu/bdd/pages/BasePage.java (hunk 5 above)
Comment (AARaven) [reply src/test/java/com/sperasoft/edu/bdd/pages/BasePage.java]: What do you mean?Could you please tell me about this directly?

File: src/main/java/com/sperasoft/edu/bdd/page/AbstractWebPage.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+public abstract class AbstractWebPage implements WebPage {
+
+    private static final String WEBPAGE_PROPERTIES_PATH =
+            "./src/main/resources/properties/webpages.properties";
+
+    @Getter
+    private URL url;
+
+    @Getter ( AccessLevel.PROTECTED )
```
Comment (a-klimov) [inline src/main/java/com/sperasoft/edu/bdd/page/AbstractWebPage.java]: What is the purpose of protected here?

File: src/main/java/com/sperasoft/edu/bdd/page/WebPage.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package com.sperasoft.edu.bdd.page;
+
+public interface WebPage {
```
Comment (a-klimov) [inline src/main/java/com/sperasoft/edu/bdd/page/WebPage.java]: From my assumption, we are not going to have more than one implementation for this interface. It looks redundant to me

File: src/main/java/com/sperasoft/edu/bdd/page/AbstractWebPage.java (hunk 10 above)
Comment (AARaven) [reply src/main/java/com/sperasoft/edu/bdd/page/AbstractWebPage.java]: I use this, ' cause if we implement some URL initialization method,we may encounter a throw MelformedUrlException, so I hide access to this field and initialize it with a hard link to the webpages file.properties

File: src/main/java/com/sperasoft/edu/bdd/page/WebPage.java (hunk 11 above)
Comment (AARaven) [reply src/main/java/com/sperasoft/edu/bdd/page/WebPage.java]: I agree, but with the expansion of the project structure, we might be faced with the need to use further abstracted methods and factories. ( GoF Design Patterns )

**Pull request merged**